### PR TITLE
$stripe->rawRequest 

### PR DESCRIPTION
**Pull request by richardm-stripe**

### CODE REVIEW

File: lib/BaseStripeClientInterface.php [hunk 1]
```diff
@@ -41,4 +41,27 @@ public function getConnectBase();
...
+     * covered yet in stripe-php.
+     *
+     * @param 'delete'|'get'|'post' $method the HTTP method
+     * @param string $path the path of the request
+     * @param array $params the parameters of the request
+     * @param array $opts the special modifiers of the request
+     *
+     * @return \Stripe\ApiResponse
+     */
+    public function rawRequest($method, $path, $params, $opts);
```
Comment (pakrym-stripe) [inline lib/BaseStripeClientInterface.php:57]: Same as with .NET let's keep these new method off the interface for now.

File: lib/Preview.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    protected $client;
+
+    public function __construct($client)
+    {
+        $this->client = $client;
+    }
+
+    private function getDefaultOpts($opts)
+    {
+        return \array_merge(['api_mode' => 'preview', 'stripe_version' => \Stripe\Util\ApiVersion::PREVIEW], $opts);
```
Comment (pakrym-stripe) [inline lib/Preview.php:19]: Shouldn't need 'stripe_version' => \Stripe\Util\ApiVersion::PREVIEW

File: tests/Stripe/BaseStripeClientTest.php [hunk 3]
```diff
@@ -202,4 +207,210 @@ public function testRequestWithOptsInParamsWarns()
...
+
+    public function testJsonRawRequestGetWithURLParams()
+    {
+        $curlClientStub = $this->getMockBuilder(\Stripe\HttpClient\CurlClient::class)
+            ->setMethods(['executeRequestWithRetries'])
+            ->getMock()
+        ;
+        $curlClientStub->method('executeRequestWithRetries')
+            ->willReturn(['{}', 200, []])
+        ;
```
Comment (pakrym-stripe) [inline tests/Stripe/BaseStripeClientTest.php:219]: nit: interesting formatting.

File: tests/Stripe/BaseStripeClientTest.php [hunk 4]
```diff
@@ -202,4 +207,210 @@ public function testRequestWithOptsInParamsWarns()
...
+        static::assertArrayNotHasKey(\CURLOPT_POSTFIELDS, $opts);
+        $content_type = null;
+        foreach ($opts[\CURLOPT_HTTPHEADER] as $header) {
+            if (self::headerStartsWith($header, 'Content-Type:')) {
+                $content_type = $header;
+            }
+        }
+        // The library sends Content-Type even with no body, so assert this
+        // But it would be more correct to not send Content-Type
+        static::assertStringStartsWith('Content-Type: application/x-www-form-urlencoded', $content_type);
```
Comment (pakrym-stripe) [inline tests/Stripe/BaseStripeClientTest.php:248]: Why starts with and not equals?

File: tests/Stripe/BaseStripeClientTest.php [hunk 5]
```diff
@@ -202,4 +207,210 @@ public function testRequestWithOptsInParamsWarns()
...
+        $stripe_version = null;
+        foreach ($opts[\CURLOPT_HTTPHEADER] as $header) {
+            if (self::headerStartsWith($header, 'Content-Type:')) {
+                $content_type = $header;
+            }
+            if (self::headerStartsWith($header, 'Stripe-Version:')) {
+                $stripe_version = $header;
+            }
+        }
+        // The library sends Content-Type even with no body, so assert this
```
Comment (pakrym-stripe) [inline tests/Stripe/BaseStripeClientTest.php:411]: Is the service OK with this?

**Review by pakrym-stripe — APPROVED** [review]
Mostly nits

File: tests/Stripe/BaseStripeClientTest.php (hunk 5 above)
Comment (anniel-stripe) [reply tests/Stripe/BaseStripeClientTest.php:411]: Yep, this is something this library and a few of the other SDKs have been doing. We should change this though.

File: tests/Stripe/BaseStripeClientTest.php (hunk 3 above)
Comment (anniel-stripe) [reply tests/Stripe/BaseStripeClientTest.php]: Oh weird, I guess it's a `PHP-CS-Fixer` thing. Same formatting in [StripeTelemetryTest](https://github.com/stripe/stripe-php/blob/6357a735/tests/Stripe/StripeTelemetryTest.php#L43)

**Pull request merged**